### PR TITLE
nsis: allow the installer to run on non-admin users

### PR DESCRIPTION
**Pull request by Lucarda**
this PR is targeted to `develop` branch.

this rewrite handles 'default writable target dirs' and 'registry settings' depending if the user is admin or not. 

if it's an admin the installation is done for all users.

### CODE REVIEW

**Comment by Lucarda** [comment]
@umlaeute here is an artifact to test:

https://nc.nubegris.com.ar/index.php/s/jFwoCmZrbEFjK2n

everything is working Ok and faster (because the older installer printed to the console each file that gets written).

The only minor issue is that we don't get the app listed in the "new" part of the win11 start-menu. (the folder and shortcuts get created)

**Comment by Lucarda** [comment]
here is an artifact to test (non-admin users + icons&art):

https://nc.nubegris.com.ar/index.php/s/b6pXiy7j62oWMs4

**Comment by umlaeute** [comment]
i haven't checked your builds yet, but my local (cross-build) ones look cool... nice!

> microsoft recommends to not provide start-menu shortcuts for un-installers

fair enough.

> The only minor issue is that we don't get the app listed in the "new" part of the win11 start-menu.

any idea what causes this?
or is does this only happen when installing as an ordinary user (as opposed to an admin), as discussed in #1904?

**Comment by Lucarda** [comment]
> any idea what causes this?

not really. I could hack it via d462142 by redefining `SetShellVarContext` in the startmenu section.

I think this PR is done and I'm un-drafting it. I think is Ok to go (i had already tested it enough).

File: msw/build-nsi.sh [hunk 1]
```diff
@@ -190,11 +190,11 @@ case $nsis_exit in
...
-cp "${SCRIPTDIR}/installer-art/big.bmp" ${OUTDIR}/big.bmp
-cp "${SCRIPTDIR}/installer-art/small.bmp" ${OUTDIR}/small.bmp
-cp "${SCRIPTDIR}/installer-art/small-un.bmp" ${OUTDIR}/small-un.bmp
-cp "${SCRIPTDIR}/installer-art/pd-un.ico" ${OUTDIR}/pd-un.ico
-cp "${SCRIPTDIR}/../tcl/pd.ico" ${OUTDIR}/pd.ico
+cp ${SCRIPTDIR}/installer-art/big.bmp ${OUTDIR}/big.bmp
+cp ${SCRIPTDIR}/installer-art/small.bmp ${OUTDIR}/small.bmp
+cp ${SCRIPTDIR}/installer-art/small-un.bmp ${OUTDIR}/small-un.bmp
+cp ${SCRIPTDIR}/installer-art/pd-un.ico ${OUTDIR}/pd-un.ico
+cp ${SCRIPTDIR}/../tcl/pd.ico ${OUTDIR}/pd.ico
```
Comment (umlaeute) [inline msw/build-nsi.sh:197]: that doesn't look right.
why are the quotes removed? (was that me??)
there should *also* be quotes around the `${OUTDIR}/...`

File: msw/pd.nsi [hunk 2]
```diff
@@ -144,6 +147,12 @@ SectionGroup /e "${PRODUCT_NAME}"
     SectionEnd
 
     Section "Create Startmenu entry" StartMenu
+        ; ugly hack so the app shows up in "recently added"
+        ${If} $CONTEXT == "all"
+            SetShellVarContext all
+        ${Else}
+            SetShellVarContext current
```
Comment (umlaeute) [inline msw/pd.nsi:154]: can't we use `SetShellVarContext $CONTEXT` (also in L60 resp L65)

File: msw/build-nsi.sh (hunk 1 above)
Comment (Lucarda) [reply msw/build-nsi.sh:197]: i did a `force push` that didn't change anything (my bad) and 1minute after you `force push`ed the fix. while I was trying to fix my bad I got lost in history. I accidentally tried to fix conflicts but something went wrong.

File: msw/pd.nsi (hunk 2 above)
Comment (Lucarda) [reply msw/pd.nsi:154]: I did try to use `SetShellVarContext $CONTEXT` (that was the initial intencion) but it didn't worked for me. I might have it wrong but it looks the compiler does not allow the expansion there? 

will investigate.

Comment (Lucarda) [reply msw/pd.nsi:154]: not sure why but we can't:

```
Usage: SetShellVarContext all|current
Error in script "C:/msys64/tmp/tmp.qMn9ZCGmH5/pd.nsi" on line 156 -- aborting creation process
```

if i pop a messegebox i know `$CONTEXT` expands to `all`. the msgbox pops up almost at the end of the "copy files" part.

**Comment by umlaeute** [comment]
Don't worry. NSIS syntax is arcane...

File: msw/pd.nsi (hunk 2 above)
Comment (Lucarda) [reply msw/pd.nsi:154]: > Don't worry. NSIS syntax is arcane...

good.

File: msw/build-nsi.sh (hunk 1 above)
Comment (Lucarda) [reply msw/build-nsi.sh]: now fixed via 687c970

**Comment by umlaeute** [comment]
i did a quick test on a VM, and it looks very cool.

two tiny remarks:
- when installing as a unprivileged user, I was nevertheless asked for an admin password (after giving the admin password, the installation proceeded to per-user setup).
  the admin password popup appeared to be in the context of "Unsafe application downloaded from the internet", so this might just be another stab of M$ to prevent innocent users from installing random malware. is this normal?
-  the icon with the red toggle for uninstalling is nice, though it raised a slight irritation in me (but nothing i couldn't live with)

and since we are there:
some other applications show the "Estimated installed size" in the `Add/Remove Programs` applet.
should we add something like this as well?

it's really lowest priority, but just in case: https://superuser.com/questions/570929/

**Comment by Lucarda** [comment]
> when installing as a unprivileged user, I was nevertheless asked for an admin password (after giving the admin password, the installation proceeded to per-user setup).
the admin password popup appeared to be in the context of "Unsafe application downloaded from the internet", so this might just be another stab of M$ to prevent innocent users from installing random malware. is this normal?

I don't understand why is this happening. on my win11 *pro* dev machine i have


|user   |with pass  |prompts for password   |installs to|
|:---   |:---:      |:---:                  | ---:      |
|admin  |no pass    |yes                    |all users  |
|user1  |no pass    |no                     |user       |
|user2  |pass       |no                     |user       |

and all works as expected.

note that same file downloaded from the internet on the "admin" user the icon gets an added shield:

nsis-shield

while on the other unprivileged users not

nsis-no-shield



- Are you using a Win *home* edition?

- does your non-privileged user has a password?
  - what happens if you provide the *user* pass instead of the *admin* pass

Note that at the nsis code level there's nothing else to do (i think).

**Comment by umlaeute** [comment]
- OS: `Windows 10 Enterprise Evaluation`
- the non-privilged user has a password (the admin has a password as well)
- before the install, i downloaded Pd as the user who was to perform the installation.

(in the meantime i have reset the VM, so I cannot check for any icon shield)

**Comment by Lucarda** [comment]
there might be some peculiarity I don't know on *enterprise*.

not sure if this helps: https://learn.microsoft.com/en-us/windows/security/identity-protection/user-account-control/how-user-account-control-works

**Comment by umlaeute** [comment]
This is what I get when running as the ordinary (non-admin, but password-protected) user:

smartscreen

i get the same when running as an ordinary (non-admin) *password-less* user.

there's no way to get past this screen without entering the password of an administrator.

i figure that the [SmartScreen](https://en.wikipedia.org/wiki/Microsoft_SmartScreen#SmartScreen_in_Windows) can be configured to be less strict, and my Win10/Enterprise probably has strict settings.

reasons that i could think of why you are not seeing this:
- Win11/Pro has less strict settings
- you manually relaxed your SmartScreen settings
- you build the installer on the computer you are testing (and Windows recognized the file as "produced here" and thus whitelisted it)


in any case: after allowing the installation with the admin password, the binary is marked as "good", and afterwards I can run the installer again (either as a normal user or as admin) without being prompted any more.

**Comment by umlaeute** [comment]
and here's *yet another* request (i do appreciate everything you put into this!).

why does the installer-art has this weird *brownish* color-gradient?
I don't remember having seen this color in the context of Pd ever.
could we pick a more Pdish color? (e.g. tcl-grey; or one of those pesky pastell colors that iemguis love so much; or just plain white without a gradient?)

**Comment by Lucarda** [comment]
> why does the installer-art has this weird brownish color-gradient?

now using grey background.

preview: https://nc.nubegris.com.ar/index.php/s/EgsHCQFYQnSKxng

**Comment by Lucarda** [comment]
here is an artifact for 671f385 :https://nc.nubegris.com.ar/index.php/s/pzRNEdjnRx9MGsn

i think this PR is done so i'm un-drafting it.

**Comment by Lucarda** [comment]
artifact for cf1d133 : https://nc.nubegris.com.ar/index.php/s/M4M9xTF3Dd4aZj9

**Comment by Lucarda** [comment]
artifact for a5f4385 : https://nc.nubegris.com.ar/index.php/s/npLTNp7jPS9bA6D

**Comment by Lucarda** [comment]
@umlaeute i think this is ready to go.

When 0.54 is out i would like to add it to [winget]( https://learn.microsoft.com/en-us/windows/package-manager/winget/) the Windows package manager (works out-of-the-box on Win11 or on up-to-date-Win10). So you can do on a terminal:

`winget install puredata`

if you do now `winget search puredata` you get 0.51. Someone has uploded the `Manifest`: 

https://github.com/microsoft/winget-pkgs/tree/master/manifests/m/MillerPuckette/PureData

Anyway: is there something else to be done before you merge?

**Comment by umlaeute** [comment]
@Lucarda now that the branch has been merged, you could delete it from your repository :-)
https://github.com/lucarda/pure-data/branches

**Comment by Lucarda** [comment]
done

**Pull request merged**